File: fastapi-backend/app/schemas/inventory.py
"""
Inventory Schemas
Pydantic schemas for inventory validation and response
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, Literal
from datetime import datetime, timedelta
from app.models.stock_unit import StockUnit


class InventoryBase(BaseModel):
    """Base inventory item schema"""
    name: str = Field(..., min_length=1, max_length=255, description="Item name")
    description: Optional[str] = Field(None, max_length=1000, description="Item description")
    category_id: Optional[int] = Field(None, description="Category ID")

    # Stock management fields
    stock_quantity: int = Field(default=0, ge=0, description="Current stock quantity")
    stock_unit: str = Field(default="pcs", description="Unit of measurement (pcs, l, kg, ml, g, etc.)")
    low_stock_threshold: int = Field(default=10, ge=0, description="Low stock alert threshold")
    reorder_level: int = Field(default=5, ge=0, description="Reorder point")
    reorder_quantity: int = Field(default=50, ge=1, description="Suggested restock quantity")

    # Container tracking (optional)
    container_type: Optional[str] = Field(None, max_length=50, description="Container type (Box, Bottle, Sack, etc.)")
    container_capacity: Optional[int] = Field(None, ge=1, description="Container capacity in stock_unit")

    @field_validator('stock_unit')
    @classmethod
    def validate_stock_unit(cls, v):
        """Validate that the stock unit is valid"""
        valid_units = [u['value'] for u in StockUnit.get_all_units()]
        if v not in valid_units:
            raise ValueError(f"Invalid stock unit. Must be one of: {', '.join(valid_units)}")
        return v


class InventoryCreate(InventoryBase):
    """Schema for creating an inventory item"""
    pass


class InventoryUpdate(BaseModel):
    """Schema for updating an inventory item"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    category_id: Optional[int] = None

    # Stock management fields (optional for update)
    stock_quantity: Optional[int] = Field(None, ge=0)
    stock_unit: Optional[str] = Field(None, description="Unit of measurement")
    low_stock_threshold: Optional[int] = Field(None, ge=0)
    reorder_level: Optional[int] = Field(None, ge=0)
    reorder_quantity: Optional[int] = Field(None, ge=1)

    # Container tracking (optional)
    container_type: Optional[str] = Field(None, max_length=50)
    container_capacity: Optional[int] = Field(None, ge=1)

    @field_validator('stock_unit')
    @classmethod
    def validate_stock_unit(cls, v):
        """Validate that the stock unit is valid"""
        if v is not None:
            valid_units = [u['value'] for u in StockUnit.get_all_units()]
            if v not in valid_units:
                raise ValueError(f"Invalid stock unit. Must be one of: {', '.join(valid_units)}")
        return v


class StockLevel(BaseModel):
    """Stock level status for inventory display"""
    current: int = Field(..., description="Current stock quantity")
    unit: str = Field(..., description="Unit of measurement")
    unit_display: str = Field(..., description="Human-readable unit name")
    status: Literal["in_stock", "low_stock", "out_of_stock"] = Field(..., description="Stock status")
    threshold: int = Field(..., description="Low stock threshold")
    percentage: int = Field(..., description="Stock level percentage for progress bars")


class InventoryResponse(InventoryBase):
    """Schema for inventory item response"""
    id: int
    last_restocked_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    category: Optional[dict] = None

    # Computed fields
    stock_status: Literal["in_stock", "low_stock", "out_of_stock"]
    needs_reorder: bool
    stock_level_percentage: int
    unit_display: str
    unit_type: str
    days_since_restock: Optional[int] = None

    # Container tracking (optional)
    container_type: Optional[str] = None
    container_capacity: Optional[int] = None
    container_count: Optional[float] = None

    class Config:
        from_attributes = True

    @classmethod
    def from_orm_with_relations(cls, obj, category=None):
        """Create response with relations"""
        data = cls.from_orm(obj)

        # Add category info
        if category:
            data.category = {
                "id": category.id,
                "name": category.name
            }

        # Add computed fields
        data.stock_status = obj.stock_status
        data.needs_reorder = obj.needs_reorder
        data.stock_level_percentage = obj.stock_level_percentage
        data.unit_display = obj.get_unit_display_name()
        data.unit_type = obj.get_unit_type()

        # Calculate days since restock
        if obj.last_restocked_at:
            data.days_since_restock = (datetime.now() - obj.last_restocked_at.replace(tzinfo=None)).days

        return data


class InventoryListResponse(BaseModel):
    """Schema for list of inventory items"""
    items: list[InventoryResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class InventorySearchParams(BaseModel):
    """Schema for inventory search parameters"""
    search: Optional[str] = Field(None, description="Search in name and description")
    category_id: Optional[int] = Field(None, description="Filter by category")
    stock_status: Optional[Literal["in_stock", "low_stock", "out_of_stock"]] = Field(None, description="Filter by stock status")
    unit_type: Optional[Literal["count", "volume", "weight"]] = Field(None, description="Filter by unit type")
    page: int = Field(1, ge=1, description="Page number")
    page_size: int = Field(20, ge=1, le=100, description="Items per page")


class LowStockItemResponse(InventoryResponse):
    """Extended response for low stock items"""
    pass


class LowStockReport(BaseModel):
    """Low stock items report"""
    items: list[LowStockItemResponse]
    total_count: int
    categories: dict[str, int] = Field(default_factory=dict, description="Count by category")


class StockAdjustment(BaseModel):
    """Stock adjustment request"""
    quantity: int = Field(..., description="Quantity to add (positive) or remove (negative)")
    reason: str = Field(..., min_length=1, max_length=255, description="Reason for adjustment")
    reference: Optional[str] = Field(None, max_length=100, description="Reference number (order, invoice, etc.)")


class StockAdjustmentReason(str):
    """Predefined stock adjustment reasons"""
    SALE = "sale"               # Sold to customer
    RESTOCK = "restock"         # Restocked from supplier
    DAMAGE = "damage"           # Item damaged
    SPOILAGE = "spoilage"       # Item spoiled/expired
    THEFT = "theft"             # Item stolen
    ADJUSTMENT = "adjustment"   # Manual adjustment
    RETURN = "return"           # Customer return
    TRANSFER = "transfer"       # Transferred to/from another location
    WASTE = "waste"             # Waste/spillage


class BulkRestockRequest(BaseModel):
    """Request for bulk restocking multiple items"""
    item_ids: list[int] = Field(..., min_items=1, description="List of item IDs to restock")
    quantity: Optional[int] = Field(None, ge=1, description="Custom quantity (default: use reorder_quantity)")


class InventoryStats(BaseModel):
    """Inventory statistics for dashboard"""
    total_items: int = Field(..., description="Total number of inventory items")
    in_stock_count: int = Field(..., description="Number of items in stock")
    low_stock_count: int = Field(..., description="Number of items at low stock")
    out_of_stock_count: int = Field(..., description="Number of items out of stock")
    total_value: float = Field(..., description="Total value of inventory in pesos")
    categories: list[dict] = Field(default_factory=list, description="Breakdown by category")


class StockLevelUpdate(BaseModel):
    """Request to update stock thresholds"""
    low_stock_threshold: Optional[int] = Field(None, ge=0, description="New low stock threshold")
    reorder_level: Optional[int] = Field(None, ge=0, description="New reorder level")
    reorder_quantity: Optional[int] = Field(None, ge=1, description="New reorder quantity")

File: quickserve-frontend/types/inventory.ts
/**
 * Inventory Types
 * Type definitions for inventory management system
 */

// ============================================================================
// Stock Unit Types
// ============================================================================

/**
 * Stock unit types for inventory tracking
 * - Count units: For discrete items (pieces, packs, boxes, dozen)
 * - Volume units: For liquids (milliliters, liters, gallons, fluid ounces)
 * - Weight units: For solids (grams, kilograms, ounces, pounds)
 */
export type StockUnit =
  // Count units (discrete items)
  | 'pcs' | 'pack' | 'box' | 'dozen'
  // Volume units (liquids)
  | 'ml' | 'l' | 'gal' | 'oz_fl'
  // Weight units (solids)
  | 'g' | 'kg' | 'oz' | 'lb'

/**
 * Container type for packaging tracking
 */
export type ContainerType =
  | 'Box' | 'Bottle' | 'Sack' | 'Bag' | 'Case'
  | 'Crate' | 'Drum' | 'Jar' | 'Can' | 'Pouch'

/**
 * Unit type categories
 */
export type UnitType = 'count' | 'volume' | 'weight'

/**
 * Stock status levels
 */
export type StockStatus = 'in_stock' | 'low_stock' | 'out_of_stock'

// ============================================================================
// Stock Level Interface
// ============================================================================

/**
 * Stock level information for display
 */
export interface StockLevel {
  /** Current stock quantity */
  current: number
  /** Unit of measurement */
  unit: StockUnit
  /** Human-readable unit name */
  unit_display: string
  /** Stock status (in_stock, low_stock, out_of_stock) */
  status: StockStatus
  /** Low stock threshold */
  threshold: number
  /** Stock level as percentage (for progress bars) */
  percentage: number
}

// ============================================================================
// Inventory Item Interface
// ============================================================================

/**
 * Inventory item with full stock information
 */
export interface InventoryItem {
  id: number
  name: string
  description?: string
  category?: {
    id: number
    name: string
  }
  created_at?: string
  updated_at?: string

  // Inventory fields
  stock_quantity: number
  stock_unit: StockUnit
  low_stock_threshold: number
  reorder_level: number
  reorder_quantity: number
  last_restocked_at?: string

  // Computed fields
  stock_status: StockStatus
  stock_level_percentage: number
  unit_display: string
  unit_type: UnitType
  needs_reorder: boolean
  days_since_restock?: number

  // Container tracking (optional)
  container_type?: ContainerType | null
  container_capacity?: number | null
  container_count?: number | null
}

// ============================================================================
// Inventory Stats Interface
// ============================================================================

/**
 * Inventory statistics for dashboard
 */
export interface InventoryStats {
  total_items: number
  in_stock_count: number
  low_stock_count: number
  out_of_stock_count: number
  total_value: number
  categories: {
    name: string
    count: number
  }[]
}

// ============================================================================
// Stock Adjustment Interface
// ============================================================================

/**
 * Stock adjustment request
 */
export interface StockAdjustment {
  /** Quantity to add (positive) or remove (negative) */
  quantity: number
  /** Reason for adjustment */
  reason: string
  /** Optional reference number (order ID, invoice, etc.) */
  reference?: string
}

/**
 * Predefined stock adjustment reasons
 */
export const STOCK_ADJUSTMENT_REASONS = [
  { value: 'sale', label: 'Sale', description: 'Sold to customer' },
  { value: 'restock', label: 'Restock', description: 'Restocked from supplier' },
  { value: 'damage', label: 'Damage', description: 'Item damaged' },
  { value: 'spoilage', label: 'Spoilage', description: 'Item spoiled/expired' },
  { value: 'theft', label: 'Theft', description: 'Item stolen' },
  { value: 'adjustment', label: 'Adjustment', description: 'Manual adjustment' },
  { value: 'return', label: 'Return', description: 'Customer return' },
  { value: 'transfer', label: 'Transfer', description: 'Transferred to/from another location' },
  { value: 'waste', label: 'Waste', description: 'Waste/spillage' },
] as const

export type StockAdjustmentReason = typeof STOCK_ADJUSTMENT_REASONS[number]['value']

// ============================================================================
// Inventory Filters Interface
// ============================================================================

/**
 * Inventory filter options
 */
export interface InventoryFilters {
  /** Search query */
  search: string
  /** Category filter */
  category: string
  /** Stock status filter */
  stockStatus: StockStatus | 'all'
  /** Unit type filter */
  unitType: UnitType | 'all'
  /** Container tracking filter */
  containerTracking: 'all' | 'with_containers' | 'without_containers'
}

// ============================================================================
// API Response Interfaces
// ============================================================================

/**
 * Response from inventory items list endpoint
 */
export interface InventoryListResponse {
  items: InventoryItem[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

/**
 * Low stock report response
 */
export interface LowStockReport {
  items: InventoryItem[]
  total_count: number
  categories: Record<string, number>
}

/**
 * Bulk restock response
 */
export interface BulkRestockResponse {
  results: Record<number, boolean>
  summary: {
    total: number
    success: number
    failed: number
  }
}

// ============================================================================
// Create Inventory Item Interface
// ============================================================================

/**
 * Input for creating a new inventory item
 */
export interface CreateInventoryItemInput {
  name: string
  description?: string
  category_id?: number | null
  // Inventory fields
  stock_quantity: number
  stock_unit: StockUnit
  low_stock_threshold: number
  reorder_level: number
  reorder_quantity: number
  // Container tracking (optional)
  container_type?: ContainerType | null
  container_capacity?: number | null
}

/**
 * Input for updating an inventory item
 */
export interface UpdateInventoryItemInput {
  name?: string
  description?: string
  category_id?: number | null
  stock_quantity?: number
  stock_unit?: StockUnit
  low_stock_threshold?: number
  reorder_level?: number
  reorder_quantity?: number
  // Container tracking (optional)
  container_type?: ContainerType | null
  container_capacity?: number | null
}

// ============================================================================
// Stock Unit Display Info
// ============================================================================

/**
 * Stock unit metadata for display
 */
export interface StockUnitInfo {
  value: StockUnit
  label: string
  type: UnitType
  icon: string
}

/**
 * Stock units organized by type
 */
export const STOCK_UNITS_BY_TYPE: Record<UnitType, StockUnitInfo[]> = {
  count: [
    { value: 'pcs', label: 'Pieces', type: 'count', icon: '📦' },
    { value: 'pack', label: 'Pack', type: 'count', icon: '📦' },
    { value: 'box', label: 'Box', type: 'count', icon: '📦' },
    { value: 'dozen', label: 'Dozen', type: 'count', icon: '📦' },
  ],
  volume: [
    { value: 'ml', label: 'Milliliters', type: 'volume', icon: '💧' },
    { value: 'l', label: 'Liters', type: 'volume', icon: '💧' },
    { value: 'gal', label: 'Gallons', type: 'volume', icon: '💧' },
    { value: 'oz_fl', label: 'Fl. Oz', type: 'volume', icon: '💧' },
  ],
  weight: [
    { value: 'g', label: 'Grams', type: 'weight', icon: '⚖️' },
    { value: 'kg', label: 'Kilograms', type: 'weight', icon: '⚖️' },
    { value: 'oz', label: 'Ounces', type: 'weight', icon: '⚖️' },
    { value: 'lb', label: 'Pounds', type: 'weight', icon: '⚖️' },
  ],
}

/**
 * Get stock unit info by value
 */
export function getStockUnitInfo(unit: StockUnit): StockUnitInfo | undefined {
  for (const typeUnits of Object.values(STOCK_UNITS_BY_TYPE)) {
    const found = typeUnits.find(u => u.value === unit)
    if (found) return found
  }
  return undefined
}

/**
 * Get all stock units as flat array
 */
export function getAllStockUnits(): StockUnitInfo[] {
  return Object.values(STOCK_UNITS_BY_TYPE).flat()
}

/**
 * Get units by type
 */
export function getUnitsByType(type: UnitType): StockUnitInfo[] {
  return STOCK_UNITS_BY_TYPE[type] || []
}

/**
 * Get display name for unit
 */
export function getUnitDisplay(unit: StockUnit): string {
  const info = getStockUnitInfo(unit)
  return info?.label || unit
}

/**
 * Get unit type
 */
export function getUnitType(unit: StockUnit): UnitType {
  const info = getStockUnitInfo(unit)
  return info?.type || 'count'
}

/**
 * Get unit icon
 */
export function getUnitIcon(unit: StockUnit): string {
  const info = getStockUnitInfo(unit)
  return info?.icon || '📦'
}

/**
 * Format stock quantity with unit
 * Example: "12 L", "150 pcs", "2.5 kg"
 */
export function formatStockQuantity(quantity: number, unit: StockUnit): string {
  const display = getUnitDisplay(unit)
  return `${quantity} ${display}`
}

/**
 * Format currency for inventory values
 */
export function formatInventoryValue(value: number): string {
  return `₱${value.toLocaleString('en-PH', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`
}

/**
 * Get stock status config for badges
 */
export const STOCK_STATUS_CONFIG: Record<StockStatus, {
  label: string
  color: string
  bg: string
  icon: string
}> = {
  in_stock: {
    label: 'In Stock',
    color: 'text-green-600',
    bg: 'bg-green-100',
    icon: '✓',
  },
  low_stock: {
    label: 'Low Stock',
    color: 'text-orange-600',
    bg: 'bg-orange-100',
    icon: '⚠',
  },
  out_of_stock: {
    label: 'Out of Stock',
    color: 'text-red-600',
    bg: 'bg-red-100',
    icon: '✗',
  },
}

/**
 * Get stock status config
 */
export function getStockStatusConfig(status: StockStatus) {
  return STOCK_STATUS_CONFIG[status]
}
